contracts/src/libraries/LPMath.sol

```diff
@@ -50,18 +54,72 @@ library LPMath {
             int256 shareAdjustment,
             uint256 bondReserves
         )
+    {
+        bool success;
+        (
+            shareReserves,
+            shareAdjustment,
+            bondReserves,
+            success
+        ) = calculateUpdateLiquiditySafe(
+            _shareReserves,
+            _shareAdjustment,
+            _bondReserves,
+            _minimumShareReserves,
+            _shareReservesDelta
+        );
+        if (!success) {
+            revert IHyperdrive.UpdateLiquidityFailed();
+        }
+    }
+
+    /// @dev Calculates the new share reserves, share adjustment, and bond
+    ///      reserves after liquidity is added or removed from the pool. This
+    ///      update is made in such a way that the pool's spot price remains
+    ///      constant.
+    /// @param _shareReserves The current share reserves.
+    /// @param _shareAdjustment The current share adjustment.
+    /// @param _bondReserves The current bond reserves.
+    /// @param _minimumShareReserves The minimum share reserves.
+    /// @param _shareReservesDelta The change in share reserves.
+    /// @return shareReserves The updated share reserves.
+    /// @return shareAdjustment The updated share adjustment.
+    /// @return bondReserves The updated bond reserves.
+    /// @return success A flag indicating whether the calculation was successful.
+    function calculateUpdateLiquiditySafe(
+        uint256 _shareReserves,
+        int256 _shareAdjustment,
+        uint256 _bondReserves,
+        uint256 _minimumShareReserves,
+        int256 _shareReservesDelta
+    )
+        internal
+        pure
+        returns (
+            uint256 shareReserves,
+            int256 shareAdjustment,
+            uint256 bondReserves,
+            bool success
+        )
     {
         // If the share reserves delta is zero, we can return early since no
         // action is needed.
         if (_shareReservesDelta == 0) {
-            return (_shareReserves, _shareAdjustment, _bondReserves);
+            return (_shareReserves, _shareAdjustment, _bondReserves, true);
         }
 
         // Update the share reserves by applying the share reserves delta. We
         // ensure that our minimum share reserves invariant is still maintained.
-        int256 shareReserves_ = _shareReserves.toInt256() + _shareReservesDelta;
-        if (shareReserves_ < _minimumShareReserves.toInt256()) {
-            revert IHyperdrive.InvalidShareReserves();
+        int256 shareReserves_;
+        (shareReserves_, success) = _shareReserves.toInt256Safe();
+        if (!success) {
+            return (0, 0, 0, false);
+        }
+        shareReserves_ += _shareReservesDelta;
+        // NOTE: This is a safe cast since the minimum share reserves must be a
+        // valid uint128.
+        if (shareReserves_ < int256(_minimumShareReserves)) {
```

Comment:
understood.  normally, it would be good to write a test that verifies that this "NOTE" holds true and the code doesn't change and invalidate this, but this might be difficult

Reply:
The `NOTE` will hold true since we assert that this is a valid `uint128` in the deployer coordinator. I get what you're saying about testing it though.